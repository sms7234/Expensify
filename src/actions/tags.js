import database from '../firebase/firebase';

export const setTags = (tags) => ({
  type: "SET_TAGS",
  tags
});

export const startSetTags = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags`)
      .once('value')
      .then((snapshot) => {
        const tags = [];
        snapshot.forEach((child) => {
          tags.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setTags(tags));
      });
  };
};

export const addTag = (tag) =>({
  type: "ADD_TAG",
  tag
});

export const startAddTag = (tagData={}) => {
  return (dispatch, getState) => {
    const {
      tag='',
      description=''
    } = tagData;
    const tagItem={tag,description};
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/tags`).push(tagItem).then((ref)=>{
      dispatch(addTag({
        id: ref.key,
        ...tagItem
      }));
    });
  };
};

export const editTag = (id, updates) => ({
  type: 'EDIT_TAG',
  id,
  updates
});

export const startEditTag = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).update(updates).then(()=>{
      dispatch(editTag(id,updates));
    });
  };
};

export const removeTag = (id) => ({
  type: 'REMOVE_TAG',
  id
});

export const startRemoveTag = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).remove().then(()=>{
      dispatch(removeTag(id));
    });
  };
};
