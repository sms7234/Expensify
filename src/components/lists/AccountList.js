import React from  'react';
import {connect} from 'react-redux';
import TwoDataFieldListItem from '../items/TwoDataFieldListItem';
import selectAccounts from '../../selectors/accounts';


export const AccountList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Accounts</div>
      <div className="show-for-desktop">Accounts</div>
      <div className="show-for-desktop">Description</div>
    </div>
    <div className="list-body">
    {
      props.accounts.length === 0 ? (
        <div className="list-item--message">
          <span>No accounts to show</span>
        </div>
      ): (
        props.accounts.map((item) => {return <TwoDataFieldListItem
       key={item.id}
       id={item.id}
       value={item.account}
       description={item.description}
       location={'editAccount'}/>;})
      )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    accounts: selectAccounts(state.accounts, state.filters)
  };
};

export default connect(mapStateToProps)(AccountList);
