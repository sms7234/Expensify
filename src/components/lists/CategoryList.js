import React from  'react';
import {connect} from 'react-redux';
import CategoryListItem from '../items/CategoryListItem';
import selectCategories from '../../selectors/categories';


export const CategoryList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Categories</div>
      <div className="show-for-desktop">Categories</div>
      <div className="show-for-desktop">Description</div>
    </div>
    <div className="list-body">
    {
      props.categories.length === 0 ? (
        <div className="list-item--message">
          <span>No categories to show</span>
        </div>
      ): (
        props.categories.map((item) => {return <CategoryListItem key={item.id}{...item}/>;})
      )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    categories: selectCategories(state.categories, state.filters)
  };
};

export default connect(mapStateToProps)(CategoryList);
