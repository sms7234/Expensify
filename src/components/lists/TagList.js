import React from  'react';
import {connect} from 'react-redux';
import TwoDataFieldListItem from '../items/TwoDataFieldListItem';
import selectTags from '../../selectors/tags';


export const TagList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Tags</div>
      <div className="show-for-desktop">Tags</div>
      <div className="show-for-desktop">Description</div>
    </div>
    <div className="list-body">
    {
      props.tags.length === 0 ? (
        <div className="list-item--message">
          <span>No tags to show</span>
        </div>
      ): (
        props.tags.map((item) => {return <TwoDataFieldListItem
       key={item.id}
       id={item.id}
       value={item.tag}
       description={item.description}
       location={'editTag'}/>;})
      )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    tags: selectTags(state.tags, state.filters)
  };
};

export default connect(mapStateToProps)(TagList);
