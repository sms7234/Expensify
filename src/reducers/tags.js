export default (state=[], action) => {
  switch(action.type){
    case 'SET_TAGS':
      return action.tags;
    case 'ADD_TAG':
      return state.concat(action.tag);
    case 'EDIT_TAG':
      return state.map((item) => {
        if (item.id === action.id){
          return {
            ...item,
            ...action.updates
          }
        } else {return item}
      });
      case 'REMOVE_TAG':
        return state.filter(({id}) => id!==action.id);
    default:
      return state;
  }
}
