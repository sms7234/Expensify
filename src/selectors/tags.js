export default (tags, {tag}) => {
  return tags.filter((item) => {
    const textMatch = item.tag.toLowerCase().includes(tag.toLowerCase());
    return textMatch;
  }).sort((a,b)=> {
    if(a.tag < b.tag){
      return -1;
    } else{
      return 1;
    }
  });
};
