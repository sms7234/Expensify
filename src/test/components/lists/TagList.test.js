import React from 'react';
import {shallow} from 'enzyme';
import {TagList} from '../../../components/lists/TagList';
import tags from '../../fixtures/tags';

test('should render TagList with tags', () => {
  const wrapper = shallow(<TagList tags={tags} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render TagList with no tags', () => {
  const wrapper = shallow(<TagList tags={[]} />);
  expect(wrapper).toMatchSnapshot();
})
