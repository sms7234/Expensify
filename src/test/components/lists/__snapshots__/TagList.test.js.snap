// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TagList with no tags 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Tags
    </div>
    <div
      className="show-for-desktop"
    >
      Tags
    </div>
    <div
      className="show-for-desktop"
    >
      Description
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item--message"
    >
      <span>
        No tags to show
      </span>
    </div>
  </div>
</div>
`;

exports[`should render TagList with tags 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Tags
    </div>
    <div
      className="show-for-desktop"
    >
      Tags
    </div>
    <div
      className="show-for-desktop"
    >
      Description
    </div>
  </div>
  <div
    className="list-body"
  >
    <CategoryListItem
      description="tax docs"
      id="1"
      key="1"
      location="editTag"
      value="tax-deductible"
    />
    <CategoryListItem
      description=""
      id="2"
      key="2"
      location="editTag"
      value="warranty"
    />
    <CategoryListItem
      description="where vacation was"
      id="3"
      key="3"
      location="editTag"
      value="location - COS"
    />
  </div>
</div>
`;
